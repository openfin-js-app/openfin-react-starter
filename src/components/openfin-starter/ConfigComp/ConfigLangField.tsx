import * as React from 'react';
import { useContext, useState }from 'react';
import { ConfigContext, I18Language } from 'react-openfin';

import { makeStyles } from '@material-ui/styles';

import Button from '@material-ui/core/Button';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Typography from '@material-ui/core/Typography';

import cx from "classnames";

import { configLangFieldCompStyle as style } from '../../../assets/jss/openfin-starter';

import i18n from '../../../i18n';

import usFlag from '../../../assets/svg/nationals/united-states.svg';
import chFlag from '../../../assets/svg/nationals/china.svg';

const useStyles = makeStyles(style);

const ConfigLangField:React.FunctionComponent<{}> = (
    {}
)=>{

    const {
        config,
        actions:{
            onUpdateLangField
        }
    } = useContext(ConfigContext);

    const classes = useStyles();

    const [anchorEl, setAnchorEl] = useState<null | HTMLElement | ((element: HTMLElement) => HTMLElement)>(null);

    const getBtnContent = ()=>{

        const language:I18Language = config.application.language;

        if (language === 'zh-CN'){
            return(<React.Fragment>
                <img className={classes.flagImg} src={chFlag}/>
                <Typography variant='body1'>简体中文</Typography>
            </React.Fragment>)
        }

        return(<React.Fragment>
            <img className={classes.flagImg} src={usFlag}/>
            <Typography variant='body1'>English(US)</Typography>
        </React.Fragment>)
    };

    const handleClick = event => {
        setAnchorEl(event.currentTarget);
    };

    const handleClose = () => {
        setAnchorEl(null);
    };

    const handleLanguageChangeBtnClick = (lang:I18Language)=>()=>{
        i18n.changeLanguage(lang);
        onUpdateLangField(lang);
        setAnchorEl(null);
    }


    return (
        <div className={classes.container}>
            <Button
                aria-owns={anchorEl ? 'language-menu' : undefined}
                aria-haspopup="true"
                onClick={handleClick}
            >
                {getBtnContent()}
            </Button>
            <Menu
                id="language-menu"
                anchorEl={anchorEl}
                open={Boolean(anchorEl)}
                onClose={handleClose}
            >
                <MenuItem onClick={handleLanguageChangeBtnClick('en-US')}>
                    <img className={classes.flagImg} src={usFlag}/>
                    <Typography variant='body1'>English(US)</Typography>
                </MenuItem>
                <MenuItem onClick={handleLanguageChangeBtnClick('zh-CN')}>
                    <img className={classes.flagImg} src={chFlag}/>
                    <Typography variant='body1'>简体中文</Typography>
                </MenuItem>
            </Menu>
        </div>
    )

}

export default ConfigLangField;

