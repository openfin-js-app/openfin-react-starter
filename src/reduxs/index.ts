import {combineReducers} from 'redux';
import {IApplicationState} from './application/types';
import {IClientState} from './client/types';
import {IConfigState} from './config/types';

import application from './application/reducer';
import { defaultState as applicationDefaultState } from './application/reducer';
import client from './client/reducer';
import { defaultState as clientDefaultState } from './client/reducer';
import config from './config/reducer';
import { defaultState as configDefaultState } from './config/reducer';

export * from './application/actions';
export * from './application/types';
export * from './client/actions';
export * from './client/types';
export * from './config/actions';
export * from './config/context';
export * from './config/types';

export interface IRootState {
    application:IApplicationState,
    client:IClientState,
    config:IConfigState,
}

export const rootDefaultState = {
    application:applicationDefaultState,
    client:clientDefaultState,
    config:configDefaultState,
}

export default (parentWindowState?:IRootState)=>{
    if (parentWindowState){
        return combineReducers({
            application:application(parentWindowState.application),
            client:client(parentWindowState.client),
            config:config(parentWindowState.config),
        });

    }else{
        return combineReducers({
            application:application(),
            client:client(),
            config:config(),
        });
    }
}
