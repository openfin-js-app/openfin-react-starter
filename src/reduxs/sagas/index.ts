import { all, takeEvery, select } from 'redux-saga/effects';

import clientSaga from './client';

export function* handleLogAllActions(action) {
    const state = yield select();
    console.log(`${action.type}`,action,'state after',state);
}

export function* watchAndLogSaga(){
    yield takeEvery('*',handleLogAllActions)
}

export default function* rootSaga(){

    const sagas = [ clientSaga() ];

    if(process.env.REACT_APP_LOG_ACTION === 'true'){
        sagas.unshift(watchAndLogSaga());
    }

    yield all(sagas);

}