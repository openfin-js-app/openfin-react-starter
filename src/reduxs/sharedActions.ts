import {
    APPLICATION_SET_LOADING_MSG,
    APPLICATION_LAUNCH_NEW_WINDOW,
} from './application/actions';

import {
    CLIENT_SET_VALUE,
} from './client/actions';

const sharedActions:string[]=[
    APPLICATION_SET_LOADING_MSG,
    APPLICATION_LAUNCH_NEW_WINDOW,
    CLIENT_SET_VALUE,
];

export default sharedActions;