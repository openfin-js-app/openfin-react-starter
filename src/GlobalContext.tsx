import * as React from 'react';
import { connect } from 'react-redux'

import { I18Language, IConfigRuntimeState } from './reduxs'

import { ConfigContextProvider } from './reduxs/config/context';

interface IProps {
    config:IConfigRuntimeState,
    onToggleThemeField?:()=>void,
    onUpdateLangField?:(lang:I18Language)=>void,
}

class GlobalContext extends React.Component<IProps,{}>{

    render(){

        const {
            config,
            onToggleThemeField,
            onUpdateLangField,
        } = this.props;

        return (
            <ConfigContextProvider value={
                {
                    config,
                    actions:{
                        onToggleThemeField,
                        onUpdateLangField,
                    }
                }
            }>
                {this.props.children}
            </ConfigContextProvider>

        )
    }
}

export default GlobalContext;