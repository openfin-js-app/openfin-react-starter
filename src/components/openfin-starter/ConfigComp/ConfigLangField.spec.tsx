import * as React from 'react';
import { act } from 'react-dom/test-utils';

import { Provider } from 'react-redux';

import { createMuiTheme } from '@material-ui/core/styles';
import { ThemeProvider } from '@material-ui/styles';
import Button from '@material-ui/core/Button';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';

import { createShallow, createMount } from '@material-ui/core/test-utils';
import configurestore from 'redux-mock-store';

import ConfigLangField from './ConfigLangField';

import GlobalContext from '../../../GlobalContext';
import {I18Language} from "../../../reduxs";

const theme = createMuiTheme({});
const mockStore = configurestore();
const zhState = {
    config:{
        application:{
            language:I18Language.zh_CN
        },
    },
};
const enState = {
    config:{
        application:{
            language:I18Language.en_US
        },
    },
};

describe('ConfigLangField',()=>{

    let shallow;
    let mount;

    beforeEach(() => {
        mount = createMount();
        shallow = createShallow();
    });

    afterEach(()=>{
        mount.cleanUp();
    });

    it('render en-US correctly and switch to zh-CN',()=>{

        const store = mockStore(enState);
        const handleUpdateLangField = jest.fn();
        const wrapper = mount(
            <Provider store={store}>
                <GlobalContext
                    config={enState.config}
                    onUpdateLangField={handleUpdateLangField}
                >
                    <ThemeProvider theme={theme}>
                        <ConfigLangField/>
                    </ThemeProvider>
                </GlobalContext>
            </Provider>
        );
        const btns = wrapper.find(Button);
        expect(btns).toHaveLength(1);

        // btns.props().onClick({event:{currentTarget:btns}});
        btns.simulate('click')

        // expect(wrapper.find(MenuItem)).toHaveLength(2);

        const menu = wrapper.find(Menu);
        menu.props().onClose();

        const menuItems = wrapper.find(MenuItem);
        expect(menuItems).toHaveLength(2);

        menuItems.at(1).simulate('click');

        expect(handleUpdateLangField).toHaveBeenCalled();
        expect(handleUpdateLangField.mock.calls).toMatchSnapshot();

    })

    it('render zh-CN correctly',()=>{

        const store = mockStore(zhState);
        const handleUpdateLangField = jest.fn();

        const wrapper = mount(
            <Provider store={store}>
                <GlobalContext
                    config={enState.config}
                    onUpdateLangField={handleUpdateLangField}
                >
                    <ThemeProvider theme={theme}>
                        <ConfigLangField/>
                    </ThemeProvider>
                </GlobalContext>
            </Provider>
        );

        expect(wrapper.find(ConfigLangField)).toBeTruthy();

    })

})