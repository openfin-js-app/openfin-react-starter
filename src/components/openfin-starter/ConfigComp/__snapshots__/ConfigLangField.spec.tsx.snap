// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigLangField render en-US correctly and switch to zh-CN 1`] = `
Array [
  Array [
    "zh-CN",
  ],
]
`;
